### PR TITLE
Add VHDL Blink example

### PR DESCRIPTION
**Pull request by umarcor**
In this PR an equivalent to verilog/blink written in VHDL is contributed.

- References implying that Verilog is the only supported HDL language are modified.
- A reference to an unexisting file (`docs/spinal.rst`) is removed.
- The documentation of verlog/blink regarding signals being exposed is fixed.
- verilog/pcf is moved to the root, because it is to be shared for verilog and VHDL examples.
- vhdl/blink is added.
- Some content in the docs is moved from `verilog.rst` to `hdl.rst`.

The makefile in vhdl/blink is based on [ghdl/ghdl-yosys-plugin: examples/](https://github.com/ghdl/ghdl-yosys-plugin/tree/master/examples) and [antonblanchard/ghdl-yosys-blink](https://github.com/antonblanchard/ghdl-yosys-blink). Hence, apart from "native" execution, Docker and Podman are supported too. Since GHDL and ghdl-yosys-plugin are not included in Fomu's toolchain yet, the makefile defaults to using Docker.

However, Fomu's toolchain is based on the same origin as [open-tool-forge/fpga-toolchain](https://github.com/open-tool-forge/fpga-toolchain) (see #319) and fpga-toolchain supports GHDL already. @xobs, any thoughts?

GitHub Actions does NOT support running GNU/Linux containers on Windows or Mac OS. That's why the VHDL example is skipped on those platforms. Nevertheless, I tested it on Windows 10, using a MINGW64 (MSYS2) terminal:

```sh
git clone -b vhdl https://github.com/umarcor/fomu-workshop
cd fomu-workshop
export FOMU_REV=pvt
cd vhdl/blink
PATH=$PATH:/path/to/fomu/toolchain/bin make
```

### CODE REVIEW

File: docs/hdl.rst [hunk 1]
```diff
@@ -2,10 +2,43 @@ Hardware Description Languages
 ------------------------------
 
 The two most common **H**\ ardware **D**\ escription **L**\ anguages are
-Verilog and VHDL (the toolchain we are using only supports Verilog).
+Verilog and VHDL.
+
+.. NOTE:: The pre-built toolchain we are releasing supports Verilog only.
+  However, `GHDL <https://github.com/ghdl>`_ might be used as a VHDL frontend
```
Comment (mithro) [inline docs/hdl.rst:8]:
```suggestion
  However, `GHDL <https://github.com/ghdl>`_ can be used as a VHDL frontend
```
or
```suggestion
  However, `GHDL <https://github.com/ghdl>`_ might be usable as a VHDL frontend
```

File: docs/hdl.rst [hunk 2]
```diff
@@ -2,10 +2,43 @@ Hardware Description Languages
...
+  for Yosys. See `ghdl-yosys-plugin <https://github.com/ghdl/ghdl-yosys-plugin>`_.
+
+When writing HDL, a tool called ``yosys`` is used to convert the
```
Comment (mithro) [inline docs/hdl.rst:11]: I think a lot of this information probably belongs in a separate section?

File: docs/hdl.rst [hunk 3]
```diff
@@ -2,10 +2,43 @@ Hardware Description Languages
...
+human readable verilog into a netlist representation, this is called
+*synthesis*. Once we have the netlist representation, a tool called
+``nextpnr`` performs an operation called *place and route* (P&R) which
+makes it something that will actually run on the FPGA. This is all
+done for you using the ``Makefiles`` in the subdirectories of ``verilog``
+or ``vhdl``.
+
+A big feature of ``nextpnr`` over its predecessor, is the fact that
```
Comment (mithro) [inline docs/hdl.rst:19]: Nobody needs to know about `nextpnr`'s predecessor.

File: verilog/blink/Makefile [hunk 4]
```diff
@@ -40,7 +40,7 @@ blink.json: blink.v
 
 # Use **nextpnr** to generate the FPGA configuration.
 # This is called the **place** and **route** step.
-blink.asc: blink.json ../pcf/fomu-hacker.pcf
+blink.asc: blink.json
```
Comment (mithro) [inline verilog/blink/Makefile:43]: This should still probably depend on the pcf file?

**Review by mithro — COMMENTED** [review]
This is a great addition!

**Comment by mithro** [comment]
We have plans to move to open-tool-forge/fpga-toolchain (see #319) but just haven't done it yet....

File: docs/hdl.rst (hunk 2 above)
Comment (umarcor) [reply docs/hdl.rst:11]: Agree. However, note that all I did was move it from `verilog.rst` to `hdl.rst`. Currently, that's shown in https://workshop.fomu.im/en/latest/verilog.html

Should it be moved to https://workshop.fomu.im/en/latest/background.html#turning-code-into-gates instead?

File: docs/hdl.rst (hunk 3 above)
Comment (umarcor) [reply docs/hdl.rst:19]: Agree. I would reword it to:

> Some P&R tools, such as nextpnr, are timing-driven. This means that a design will be generated with a given clock domain guaranteed to perform fast enough, but sometimes it won't further optimise it.

Still, let's decide where to put this content first.

File: verilog/blink/Makefile (hunk 4 above)
Comment (umarcor) [reply verilog/blink/Makefile:43]: The point is that the PCF is different, depending on `FOMU_REV`. However, the PCF is not specified as a separated variable, but included in PNRFLAGS. Therefore, I think that the PCF file should be defined in a different variable, and that would be used as a dependency.

I didn't do these modifications in this PR because I wanted to "just convert the existing Verilog example". This enhancement would apply to both of them.

**Comment by umarcor** [comment]
> We have plans to move to open-tool-forge/fpga-toolchain (see #319) but just haven't done it yet....

My main concern with the current proposal is whether usage of containers might be too disruptive for the target readers of the workshop. Apart from that, it's transparent, and the workflow is exactly equivalent to the Verilog example.

Hence, I think that migrating to fpga-toolchain would be desirable for decoupling the VHDL examples from an introduction to using containers instead of native tools. Then, I would propose moving the Makefile details about how to use containers to some specific section marked as (intermediate/advanced).

**Comment by umarcor** [comment]
I updated `get-toolchain.py` for using the 'nightly' release of the "new" toolchain based on fpga-toolchain. Therefore, containers are not used anymore, and I removed any reference to Docker/Podman from this PR. All tests are now passing with the tools on the host.

**Comment by umarcor** [comment]
I had to set `GHDL_PREFIX` in `tests.sh` for ghdl-yosys-plugin to work. Otherwise, it would complain because it could not find the libraries: https://github.com/umarcor/fomu-workshop/runs/1243053586?check_suite_focus=true#step:9:47.

Locally, adding Fomu toolchain (i.e. fpga-toolchain) to the PATH does work already, without GHDL_PREFIX. But I'm afraid that's because I have another GHDL "properly" installed on the system, so it's finding `/mingw64/lib/ghdl/ieee/*`, NOT `$FOMU_TOOLCHAIN/lib/ghdl/ieee/*`.

@tgingold, should GHDL search/find libraries relative to the location of the executable? GHDL itself seems to work (`ghdl --version`, `ghdl --disp-config`, etc.), but synthesising produces an error unless GHDL_PREFIX is properly set.

@edbordin, I opened open-tool-forge/fpga-toolchain#59, but I'm not sure that will fix the problem. Did you find any similar issue? Are you aware of any user of fpga-toolchain which did actually try/use ghdl-yosys-plugin + Yosys?

**Comment by tgingold** [comment]
For ghdl-yosys-plugin (or libghdl), the library path depends on how it was configured.  It doesn't look at the ghdl executable.

**Comment by umarcor** [comment]
> For ghdl-yosys-plugin (or libghdl), the library path depends on how it was configured. It doesn't look at the ghdl executable.

Thanks! That explains the issue!

Then, I should add a clarification in the installation guidelines of this workshop. @edbordin, shall I open a PR for the README of fpga-toolchain?

**Comment by umarcor** [comment]
The requirements section of workshop is updated now.

BTW, I split the section into multiple files (`files.rst`, `software.rst`, `hardware.rst` and `drivers.rst`).

**Comment by edbordin** [comment]
> @edbordin, shall I open a PR for the README of fpga-toolchain

@umarcor I already have a note in the README about GHDL_PREFIX  if that's what you meant? Weirdly it actually works without it on windows
